### PR TITLE
Fix for server crash in multi-svr

### PR DESCRIPTION
**Pull request by nithinj**
#### Describe Bug or Feature
* Job owning server crashes when move+run job fails due to an error while submitting the job.
* Mom choosing between primary and secondary servers randomly could result in skewed results.

#### Describe Your Change
* Move and run sends two requests to the destination server - submit and run. When the submit fails due to an error, it sends the error back and the source server sends it back to the client. When it receives the error corresponding to the run request, the request is invalid as it has been already processed.
The fix is to delete any existing task corresponding to the request after replying.
* Mom will choose between primary and secondary in an interleaved way than random

#### Attach Test and Valgrind Logs/Output

[move_run_fix.log](https://github.com/openpbs/openpbs/files/5703167/move_run_fix.log)

### CODE REVIEW

File: src/resmom/catch_child.c [hunk 1]
```diff
@@ -840,13 +841,16 @@ scan_for_exiting(void)
...
-get_servername_random(unsigned int *port)
+get_servername_failover(unsigned int *port)
 {
+	static int whom_to_connect = 0;
 
 	if (!pbs_conf.pbs_secondary)
 		return get_servername(port);
 	else {
-		if (rand() % 2 == 0)
+		whom_to_connect = (whom_to_connect + 1) % 2;
```
Comment (agrawalravi90) [inline src/resmom/catch_child.c:851]: Simpler:
```suggestion
		whom_to_connect = !whom_to_connect;
```

**Review by agrawalravi90 — CHANGES_REQUESTED** [review]
Minor comment

File: src/server/multi_svr.c [hunk 2]
```diff
@@ -279,3 +279,30 @@ init_msi()
...
+char *
+get_svr_inst_id(void)
+{
+	char svr_inst_name[PBS_MAXHOSTNAME + 1];
+	unsigned int svr_inst_port;
+	char *svr_inst_id = NULL;
+
+
+	if (gethostname(svr_inst_name, PBS_MAXHOSTNAME) == 0)
+        	get_fullhostname(svr_inst_name, svr_inst_name, PBS_MAXHOSTNAME);
```
Comment (ashwathraop) [inline src/server/multi_svr.c:300]: No need to handle possible failure of get_fullhostname()? It can return -1 right?

Comment (nithinj) [reply src/server/multi_svr.c:300]: In that case, we still retain the original value

**Comment by nithinj** [comment]
@ashwathraop & @agrawalravi90 ,
Thanks for the reviews. I have addressed your comments. Please have a look.

File: src/server/multi_svr.c [hunk 3]
```diff
@@ -279,3 +279,30 @@ init_msi()
...
+/**
+ * @brief
+ * 	Used to get serverer_instance_id which is of the form server_instance_name:server_instance_port
+ * 
+ * @return char *
+ * @return NULL - failure
+ * @retval !NULL - pointer to server_instance_id
+ */
+char *
+get_svr_inst_id(void)
```
Comment (agrawalravi90) [inline src/server/multi_svr.c:292]: since this is actually allocating memory, how about 'create_svr_inst_id'/'generate_svr_inst_id' ?

**Review by agrawalravi90 — CHANGES_REQUESTED** [review]
sorry, one more minor comment

**Comment by nithinj** [comment]
@agrawalravi90 ,
I have addressed your comments. Please have a look.

**Review by agrawalravi90 — APPROVED** [review]
Looks good now, thanks

**Review by ashwathraop — APPROVED** [review]

**Pull request merged**